qf_lib/data_providers/preset_data_provider.py

```diff
@@ -153,43 +153,31 @@ def historical_price(self, tickers: Union[Ticker, Sequence[Ticker]],
         fields, got_single_field = convert_to_list(fields, PriceField)
         got_single_date = nr_of_bars == 1
 
-        data_bundle = self._data_bundle
+        data_bundle = self._data_bundle.loc[:end_date, specific_tickers, fields].dropna(DATES, how='all')
 
-        # Data aggregation (allowed only for the Intraday Data and in case if more then 1 data point is found)
-        if frequency < self._frequency and len(data_bundle[DATES]) > 0:
-            data_bundle = self._aggregate_intraday_data_helper(data_bundle, frequency, fields)
+        if self._frequency == frequency:
+            self._check_data_availibility(data_bundle, end_date, nr_of_bars, tickers)
+            data_bundle = data_bundle.isel(dates=slice(-nr_of_bars, None))
+        elif frequency < self._frequency and data_bundle.shape[0] > 0:
+            # Data aggregation will happen only for the Intraday Data and in case if more than 1 data point is found
+            bars_len_multiplier = int(
+                ceil(to_offset(frequency.to_pandas_freq()) / to_offset(self._frequency.to_pandas_freq()))
+            )
+            # optimize by selecting only part of data_bundle to be aggregated as aggregation is slow
+            bars_before_aggregation = nr_of_bars * bars_len_multiplier * 1.2
```

Comment:
Maybe we can do some additional check instead of just using 1.2? Here we still can lose some data in the end, however I don't know how slow it will be after replacing 1.2 with some other checks

Reply:
ok, I reused existing method - for start_date calculation